File: testng-core/src/test/java/test/custom/CustomAttributesTransformer.java
package test.custom;

import java.lang.annotation.Annotation;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.Objects;
import org.testng.IAnnotationTransformer;
import org.testng.annotations.CustomAttribute;
import org.testng.annotations.ITestAnnotation;

public class CustomAttributesTransformer implements IAnnotationTransformer {

  @Override
  public void transform(
      ITestAnnotation annotation, Class testClass, Constructor testConstructor, Method testMethod) {
    CustomAttribute[] attributes =
        new CustomAttribute[] {new MoreAttribute("sorrow", "Coffee", "Tea")};
    annotation.setAttributes(attributes);
  }

  public static class MoreAttribute implements CustomAttribute {

    private final String key;
    private final String[] values;

    public MoreAttribute(String key, String... values) {
      this.key = key;
      this.values = values;
    }

    @Override
    public String name() {
      return key;
    }

    @Override
    public String[] values() {
      return values;
    }

    @Override
    public Class<? extends Annotation> annotationType() {
      return CustomAttribute.class;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      MoreAttribute that = (MoreAttribute) o;
      return Objects.equals(key, that.key) && Arrays.equals(values, that.values);
    }

    @Override
    public int hashCode() {
      int result = Objects.hash(key);
      result = 31 * result + Arrays.hashCode(values);
      return result;
    }
  }
}

File: testng-core/src/test/kotlin/test/attributes/samples/custom/CustomAttributesTransformer.kt
package test.attributes.samples.custom

import java.lang.reflect.Constructor
import java.lang.reflect.Method
import org.testng.IAnnotationTransformer
import org.testng.annotations.CustomAttribute
import org.testng.annotations.ITestAnnotation
import kotlin.reflect.KCallable

class CustomAttributesTransformer : IAnnotationTransformer {

    @Override
    override fun transform(
        annotation: ITestAnnotation?,
        testClass: Class<*>?,
        testConstructor: Constructor<*>?,
        testMethod: Method?
    ) {
        val attributes = arrayOf(
            customAttribute("sorrow", arrayOf("Coffee", "Tea")))
        annotation?.attributes = attributes
    }

    // Support for creation annotation literals is still not there in Kotlin.
    // Refer to the Open Defect https://youtrack.jetbrains.com/issue/KT-25947
    // Below work-around is the only way in which we can create annotation literals in Kotlin
    // and has been borrowed from https://stackoverflow.com/q/64539969
    private fun customAttribute(name:String, values: Array<String>) : CustomAttribute{
        val constructors = CustomAttribute::class.constructors
        val callable: KCallable<*> = constructors.first()
        val param1 = callable.parameters[0]
        val param2 = callable.parameters[1]
        return callable.callBy(mapOf(param1 to name, param2 to values)) as CustomAttribute
    }

}
